components/sup/src/cli.rs

```diff
@@ -30,6 +30,67 @@ mod test {
 
     mod sup_run {
         use super::*;
+        use crate::common::cli_defaults::{
+            GOSSIP_DEFAULT_ADDR, LISTEN_CTL_DEFAULT_ADDR_STRING, LISTEN_HTTP_DEFAULT_ADDR,
+        };
+        use std::iter::FromIterator;
+        use tempfile::{NamedTempFile, TempDir};
+
+        assert_cli_cmd!(should_handle_listen_gossip_with_value,
+                        "hab-sup run --listen-gossip 1.1.1.1:1111",
+                        "LISTEN_GOSSIP" => "1.1.1.1:1111");
+
+        assert_cli_cmd!(listen_gossip_should_have_default_value,
+                        "hab-sup run",
+                        "LISTEN_GOSSIP" => (&GOSSIP_DEFAULT_ADDR));
+
+        #[test]
+        fn invalid_listen_gossip_address_should_fail() {
+            let cmd_vec =
+                Vec::from_iter("hab-sup run --listen-gossip bad-address".split_whitespace());
+            assert!(cli().get_matches_from_safe(cmd_vec).is_err());
+        }
+
+        assert_cli_cmd!(should_handle_listen_http,
+                        "hab-sup run --listen-http 1.1.1.1:1111",
+                        "LISTEN_HTTP" => "1.1.1.1:1111");
+
+        assert_cli_cmd!(listen_http_should_have_default_value,
+                        "hab-sup run",
+                        "LISTEN_HTTP" => (&LISTEN_HTTP_DEFAULT_ADDR));
+
+        #[test]
+        fn invalid_listen_http_address_should_fail() {
+            let cmd_vec =
+                Vec::from_iter("hab-sup run --listen-http bad-address".split_whitespace());
+            assert!(cli().get_matches_from_safe(cmd_vec).is_err());
+        }
+
+        assert_cli_cmd!(should_handle_http_disable,
+                        "hab-sup run --http-disable",
+                        "HTTP_DISABLE" => true);
+
+        assert_cli_cmd!(http_disable_should_have_short_flag,
+                        "hab-sup run -D",
+                        "HTTP_DISABLE" => true);
+
+        assert_cli_cmd!(should_handle_listen_ctl,
+                        "hab-sup run --listen-ctl 1.1.1.1:1111",
+                        "LISTEN_CTL" => "1.1.1.1:1111");
+
+        assert_cli_cmd!(listen_ctl_should_have_a_default_value,
+                        "hab-sup run",
+                        "LISTEN_CTL" => (&LISTEN_CTL_DEFAULT_ADDR_STRING));
+
+        #[test]
+        fn invalid_listen_ctl_address_should_fail() {
+            let cmd_vec = Vec::from_iter("hab-sup run --listen-ctl bad-address".split_whitespace());
+            assert!(cli().get_matches_from_safe(cmd_vec).is_err());
+        }
+
+        assert_cli_cmd!(should_handle_org,
+                        "hab-sup run --org monkeypants",
+                        "ORGANIZATION" => "monkeypants");
```

Comment:
Generally I would agree with you, but the primary purpose for these tests are in preparation to refactor `sub_sup_run` and remove the macro. I would prefer to have something for each option just so I can make sure that I'm getting the proper functionality and not leaving anything out when I do the refactor.

Reply:
Unfortunately, testing is no guarantee here. If you miss a thing to test or make an error in the test code, you can let defects through as well. Beware of adding tests giving you more confidence than they should: test code is code, and at the end of the day the only way we can truly have confidence in test code's correctness is through human inspection and experience running both the test code and the system under test.

I do think the general pattern of adding good tests before a refactor is a good one, but I *don't* believe in the idea of adding tests that you wouldn't otherwise want to just have around, especially since we don't have any special knowledge of the kinds of defects this sort of refactoring is likely to introduce. Perhaps as we start the macro → builder refactor, we will learn some things which inspire more targeted testing, but for now, let's add tests that bring value to the code as it is and refrain from adding tests whose correctness is harder to verify than the code they are testing.

Reply:
I agree in principle, but I disagree in this specific case. That said, I'm fine with conceding here if you feel it would be better.

Reply:
I do. We have a lot of code, and every bit that we add increases the weight of what we have to maintain. I'm not convinced these changes pay for themselves, so let's leave them out unless something changes that calculus.

Reply:
I removed the presence and short flag tests. I'm gonna leave in the conflicts_with tests for now though; those tests already helped me find a bug.

Reply:
I understand the impulse to keep tests that helped you find a bug, but I'm not certain that's the most effective way to evaluate whether it makes sense to keep a particular test in the codebase. It's great that we found this bug (and good also that it's the only instance of this particular class of bug), but for tests that we _leave in_ the codebase, I think they should satisfy one of the following criteria:

1. Validating behavior to catch a likely regression from a future change
1. Documenting desired behavior that is made clearer through example

There's certainly an argument to be made that this test satisfies (1):
```rust
        #[test]
        fn peer_watch_file_conflicts_with_peer() {
            let cmd_vec = Vec::from_iter(
                "hab-sup run --peer-watch-file foobar --peer 1.1.1.1:1111".split_whitespace(),
            );
            assert!(cli().get_matches_from_safe(cmd_vec).is_err());
        }
```
But `conflicts_with` isn't our logic, it's a part of clap. And our bug was that we weren't using it correctly. If we add a 3-line test for every time we use `conflicts_with`, we end up with a lot of test code to manage. And any time we want to change any of the places in the code that use `conflicts_with`, we also have to change the test code. But that probably won't be recognized until the test fails in CI, causing delay. There are lots of ways we can have bugs on account of misusing a library we depend on, and I don't think that adding tests for all of them works well to help address that issue. There's alway a balance between testing thoroughly to avoid the introduction of bugs, and not testing so specifically as to hamper development on a codebase we expect to change.

I think it's great we found this bug, and in that sense, I think going through writing that test had value. But I think the tests we keep should be the ones we expect to have value in the future that exceeds the cost of maintaining them. I think this test's value was already realized and now that you've fixed the bug, I think future value is maximized by keeping the set of tests a bit leaner.